### PR TITLE
refactor(v3): finish user slice (DEV-671)

### PR DESCRIPTION
**Pull request by irinaschubert**
Resolves DEV-671

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package dsp.user.sharedtestdata
+
+import dsp.valueobjects.User._
+import dsp.user.domain.User
+
+object SharedTestData {
+  val simpleGivenName1  = GivenName.make("GivenName1").fold(e => throw e.head, v => v)
```
Comment (subotic) [inline dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala:12]: Should we ever throw? I thought that a `make` returns a `Validation`?

Comment (mpro7) [reply dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala:12]: It does, but then we either return a value object or throw an error. At least this is how it's now implemented.

File: dsp-user/repo/src/test/scala/dsp/user/repo/impl/UserRepoImplSpec.scala [hunk 2]
```diff
@@ -23,65 +26,81 @@ object UserRepoImplSpec extends ZIOSpecDefault {
...
+          fails(equalTo(None))
+        )
+      } +
+      test("return success (Unit) if a username is unique") {
+        for {
+          _      <- UserRepo.storeUser(user1)
+          result <- UserRepo.checkIfUsernameExists(SharedTestData.simpleUser2.username)
+        } yield assertTrue(result == ())
+      } +
+      test("return None if an email already exists") {
```
Comment (BalduinLandolt) [inline dsp-user/repo/src/test/scala/dsp/user/repo/impl/UserRepoImplSpec.scala:75]: the description of this test is not wrong, but maybe it would be more clear how it's being tested, if the description would say that the `None` is returned in the error channel. (As you do in the tests where you write `return success (Unit)`)

**Review by BalduinLandolt — APPROVED** [review]
Over all, looks good, thanks!

Some things I commented on:
- various cosmetic/code style things... feel free to ignore
- I think in the UserDomainSpec you can simplify the tests by not wrapping everything into a for-comprehension and into ZIOs. Those tests should not need this.
- Typo in PR title scope. Please fix :)

**Comment by BalduinLandolt** [comment]
darn... you already fixed thy typo in the PR title... ignore that bullet point then! ^^

File: dsp-user/core/src/main/scala/dsp/user/domain/UserDomain.scala [hunk 3]
```diff
@@ -5,138 +5,189 @@
...
+   *  @return the updated [[User]]
    */
-  // TODO should this return a Validation[Throwable, UserId]
-  def make(): UserId = {
-    val uuid: UUID       = UUID.randomUUID()
-    val iri: Iri.UserIri = Iri.UserIri.make("http://rdfh.ch/users/" + uuid.toString)
-    new UserId(uuid, iri) {}
-  }
-}
+  def updatePassword(newPassword: PasswordHash): User =
```
Comment (mpro7) [inline dsp-user/core/src/main/scala/dsp/user/domain/UserDomain.scala:127]: To be consistent with line 124, you should use `newValue` here ;)
Also, `id` isn't param anymore in all update methods, is it?.

File: dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala [hunk 4]
```diff
@@ -4,3 +4,126 @@
...
+ */
+object UserDomainSpec extends ZIOSpecDefault {
+
+  def spec = (userDomainTests)
+
+  val userDomainTests = suite("UserDomain")(
+    test("compare two users") {
+      val user         = SharedTestData.simpleUser1
+      val userEqual    = SharedTestData.simpleUser1 // same as user, i.e. has same UserId
+      val userNotEqual = SharedTestData.simpleUser2
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala:27]: At the end name should represent kind of user we use for test, so it could be specified now or improved later.

File: dsp-shared/src/test/scala/dsp/valueobjects/UserSpec.scala [hunk 5]
```diff
@@ -245,13 +255,35 @@ object UserSpec extends ZIOSpecDefault {
...
+      assertTrue(PasswordHash.make(validPassword).toOption.get.value == validPassword)
+    },
+    test("test if two passwords match") {
+      val password         = PasswordHash.make("password1").fold(e => throw e.head, v => v)
+      val passwordEqual    = PasswordHash.make("password1").fold(e => throw e.head, v => v)
+      val passwordNotEqual = PasswordHash.make("password1").fold(e => throw e.head, v => v)
+
+      assertTrue(password.matches(password) == true) &&
+      assertTrue(password.matches(passwordEqual) == true) &&
+      assertTrue(password.matches(passwordNotEqual) == false)
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-shared/src/test/scala/dsp/valueobjects/UserSpec.scala:274]: It isn't necessary, but in terms of tests where there are usually few lines of true/false assertions, looks more readable.

File: dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala [hunk 6]
```diff
@@ -4,3 +4,126 @@
...
+        assertTrue(updatedUser.email == user.email) &&
+        assertTrue(updatedUser.givenName == user.givenName) &&
+        assertTrue(updatedUser.familyName == user.familyName) &&
+        assertTrue(updatedUser.password == user.password) &&
+        assertTrue(updatedUser.language != user.language) &&
+        assertTrue(updatedUser.status == user.status)
+    },
+    test("update the status") {
+      val user     = SharedTestData.simpleUser1
+      val newValue = UserStatus.make(false).fold(e => throw e.head, v => v)
```
[the remark this answers is not in the document]
Comment (mpro7) [reply dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala:117]: That might make sense, but names `ThrowOrGet`? Or maybe there is better way of doing it than `fold`.

File: dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala [hunk 7]
```diff
@@ -26,104 +27,219 @@ final case class UserHandler(repo: UserRepo) {
   // implement all possible requests from V2, but divide things up into smaller functions to keep it cleaner than before
```
Comment (mpro7) [inline dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala:27]: Is this comment still needed?

File: dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala [hunk 8]
```diff
@@ -26,104 +27,219 @@ final case class UserHandler(repo: UserRepo) {
...
   ): IO[DuplicateValueException, UserId] =
-    for {
-      _      <- checkUsernameTaken(username) // also lock/reserve username
-      _      <- checkEmailTaken(email) // also lock/reserve email
-      user   <- ZIO.succeed(User.make(givenName, familyName, username, email, password, language))
-      userId <- repo.storeUser(user) // we assume that this can't fail
-    } yield userId
+    (for {
+      _      <- checkIfUsernameTaken(username) // TODO reserve username
+      _      <- checkIfEmailTaken(email) // TODO reserve email
```
Comment (mpro7) [inline dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala:119]: Can these comments be addressed in this PR or moved to another task?

File: dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala [hunk 9]
```diff
@@ -26,104 +27,219 @@ final case class UserHandler(repo: UserRepo) {
...
+  /**
+   * Updates the username of a user
+   *
+   *  @param id  the user's ID
+   *  @param newValue  the new username
+   */
+  def updateUsername(id: UserId, newValue: Username): IO[RequestRejectedException, UserId] =
+    (for {
+      _ <- checkIfUsernameTaken(newValue)
+      // TODO reserve new username because it has to be unique
```
Comment (mpro7) [inline dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala:133]: Duplicate TODO, see comment above.

**Review by mpro7 — APPROVED** [review]
Looks good, but there are few things you could improve, see comments.

File: dsp-user/handler/src/test/scala/dsp/user/handler/UserHandlerSpec.scala [hunk 10]
```diff
@@ -20,77 +27,496 @@ object UserHandlerSpec extends ZIOSpecDefault {
...
   def spec = (userTests)
 
-  private val givenName  = GivenName.make("GivenName1").fold(e => throw e.head, v => v)
-  private val familyName = FamilyName.make("familyName1").fold(e => throw e.head, v => v)
-  private val username   = Username.make("username1").fold(e => throw e.head, v => v)
-  private val email      = Email.make("email1@email.com").fold(e => throw e.head, v => v)
-  private val password   = Password.make("password1").fold(e => throw e.head, v => v)
-  private val language   = LanguageCode.make("en").fold(e => throw e.head, v => v)
-
   val userTests = suite("UserHandler")(
```
[the remark this answers is not in the document]
Comment (irinaschubert) [reply dsp-user/handler/src/test/scala/dsp/user/handler/UserHandlerSpec.scala]: I reorganized it.

File: dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala (hunk 7 above)
Comment (irinaschubert) [reply dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala]: nope

File: build.sbt [hunk 11]
```diff
@@ -231,14 +231,14 @@ lazy val webapiJavaTestOptions = Seq(
...
-  )
-  .dependsOn(schemaCore, schemaRepo, schemaApi)
+// lazy val apiMain = project
+//   .in(file("dsp-api-main"))
+//   .settings(
+//     name := "dsp-api-main",
+//     libraryDependencies ++= Dependencies.dspApiMainLibraryDependencies,
+//     testFrameworks := Seq(new TestFramework("zio.test.sbt.ZTestFramework"))
+//   )
+//   .dependsOn(schemaCore, schemaRepo, schemaApi)
```
[the remark this answers is not in the document]
Comment (irinaschubert) [reply build.sbt]: I removed it now.

File: dsp-user/repo/src/main/scala/dsp/user/repo/impl/UserRepoLive.scala [hunk 12]
```diff
@@ -5,82 +5,117 @@
...
-      _ = iriOption match {
-            case None    => ZIO.succeed(()) // username or email does not exist
-            case Some(_) => ZIO.fail(None)  // username or email does exist
-          }
-    } yield ()).commit.tap(_ => ZIO.logInfo(s"Username/email '${usernameOrEmail}' was checked"))
+      emailExists <- lookupTableEmailToUuid.contains(email).commit
+      _ <- emailExists match {
+             case false => ZIO.succeed(()) // email does not exist
+             case true  => ZIO.fail(None)  // email does exist
+           }
```
[the remark this answers is not in the document]
Comment (BalduinLandolt) [reply dsp-user/repo/src/main/scala/dsp/user/repo/impl/UserRepoLive.scala:107]: I'm ok with it like this

File: dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala (hunk 1 above)
Comment (subotic) [reply dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala:12]: I hope we cleared this part yesterday (no throwing allowed in ZIO or you need to wrap it inside a `ZIO.attempt`)

File: dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (subotic) [reply dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala:144]: You shouldn't throw in the first place (but I think that was discussed yesterday). What error will the domain entity return if the update is not possible? It would probably also make sense to return a Validation. In this case, you could have both steps inside a for comprehension, which should clean it up nicely.

Here is an example of what I mean: https://zio.github.io/zio-prelude/docs/functionaldatatypes/validation#chaining-validations

After that you can transform the Validation in whatever shape is needed: https://zio.github.io/zio-prelude/docs/functionaldatatypes/validation#extracting-validation-values

It should never be necessary to use `fold`. Using `fold` is a sign, that something is not quite right.

File: dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala [hunk 13]
```diff
@@ -23,107 +24,221 @@ import java.util.UUID
...
+   * Retrieves the user by ID.
    *
    * @param id  the user's ID
    */
   def getUserById(id: UserId): IO[NotFoundException, User] =
     for {
-      user <- repo.getUserById(id).mapError(_ => NotFoundException("User not found"))
+      user <- repo
+                .getUserById(id)
+                .tap(_ => ZIO.logInfo(s"Looked up user by ID '${id}'"))
```
Comment (subotic) [inline dsp-user/handler/src/main/scala/dsp/user/handler/UserHandler.scala:43]: I would add `.tap` at the end and after the `mapError`, as it is not part of the logic, but the `mapError` is.

**Review by subotic — APPROVED** [review]

File: dsp-shared/src/main/scala/dsp/valueobjects/Id.scala [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  /**
+   * Stores the user ID, i.e. UUID and IRI of the user
+   *
+   * @param uuid the UUID of the user
+   * @param iri the IRI of the user
+   */
+  abstract case class UserId private (
+    uuid: UUID,
+    iri: Iri.UserIri
+  )
```
[the remark this answers is not in the document]
Comment (irinaschubert) [reply dsp-shared/src/main/scala/dsp/valueobjects/Id.scala:25]: What do you mean with "everywhere"? As far as I can see it is only in this location. Are there other places where it needs the `extends Id`?

File: dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (irinaschubert) [reply dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala]: I make the `User.make(...)` also return a `Validation` as you suggested and removed all the `fold(...)`s from the tests. Could you check if it is as you imagined?

File: dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala [hunk 15]
```diff
@@ -4,3 +4,161 @@
...
+      assertTrue(!user.equals(userNotEqual)) &&
+      assertTrue(user == userEqual) &&
+      assertTrue(user != userNotEqual)
+    }
+  )
+
+  private val createUserTest = suite("createUser")(
+    test("create a user") {
+      for {
+        id         <- SharedTestData.userId1.toZIO
```
Comment (subotic) [inline dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala:39]: You can do the same thing without `.toZIO` (I think). Validations behave the same way like ZIOs.

File: dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+package dsp.user.sharedtestdata
+
+import dsp.valueobjects.User._
+import dsp.user.domain.User
+import dsp.errors.BadRequestException
+import zio.prelude.Validation
+import dsp.valueobjects.Id
+
+object SharedTestData {
+  val passwordStrength = PasswordStrength.make(12)
```
Comment (subotic) [inline dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala:15]: That is a system property coming from AppConfig which is typesafe and does't need to be a value object.

**Review by subotic — APPROVED** [review]

**Review by mpro7 — APPROVED** [review]

File: dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala (hunk 15 above)
Comment (irinaschubert) [reply dsp-user/core/src/test/scala/dsp/user/domain/UserDomainSpec.scala]: The `.toZIO` seems to be needed for the test to compile. But I now wrapped the whole for comprehension into a `.toZIO` instead of repeating it several times inside the for comprehension.

File: dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala (hunk 16 above)
Comment (irinaschubert) [reply dsp-user/core/src/test/scala/dsp/user/sharedtestdata/SharedTestData.scala:15]: I wasn't sure about how this parameter should be handled in tests. So, you would just use a plain int to provide this value to the method `PasswordHash.make()`?
